### PR TITLE
added support to duckdb via ADBC

### PR DESCRIPTION
**Pull request by matreyes**
In my role as Data Engineer I've been moving out from dataframes (aka spark) as I'm building everything with DBT and BigQuery and locally with DuckDB (CLI). Using just sql to build different stages (kind of delta lake), has been a great way to facilitate the work of data scientists and for data governance. 

I can see how DuckDB could work as the Ingestion and transformation layer to expose a simplified and maybe materialized "data marts" to explorer or vegalite. This is so powerful!

Amazing work with ADBC !

### CODE REVIEW

File: lib/kino_db/connection_cell.ex [hunk 1]
```diff
@@ -383,7 +400,7 @@ defmodule KinoDB.ConnectionCell do
       Code.ensure_loaded?(Exqlite) -> "sqlite"
       Code.ensure_loaded?(ReqBigQuery) -> "bigquery"
       Code.ensure_loaded?(ReqAthena) -> "athena"
-      Code.ensure_loaded?(Adbc) -> "snowflake"
+      Code.ensure_loaded?(Adbc) -> "duckdb"
```
Comment (josevalim) [inline lib/kino_db/connection_cell.ex:403]: This is tricky. Adbc supports duckdb now, but it also snowflake and soon bigquery. I wonder which one should really be the default?

**Comment by josevalim** [comment]
Thank you @matreyes! I have some questions to understand your use case better:

1. Why not use Explorer/DataFrames for the data manipulation?
2. Do you actually have a Duck DB in disk?
3. How are you currently importing data into Duck DB to work with it?

File: lib/kino_db/connection_cell.ex (hunk 1 above)
Comment (josevalim) [reply lib/kino_db/connection_cell.ex:403]: Three good arguments. :)

File: lib/kino_db/connection_cell.ex [hunk 2]
```diff
@@ -400,7 +402,7 @@ defmodule KinoDB.ConnectionCell do
       Code.ensure_loaded?(Exqlite) -> "sqlite"
       Code.ensure_loaded?(ReqBigQuery) -> "bigquery"
       Code.ensure_loaded?(ReqAthena) -> "athena"
-      Code.ensure_loaded?(Adbc) -> "duckdb"
+      Code.ensure_loaded?(Adbc) -> "snowflake"
```
Comment (josevalim) [inline lib/kino_db/connection_cell.ex:405]: lets keep duck! you convinced me!

**Comment by matreyes** [comment]
Answering your questions

1. Why not use Explorer/DataFrames for the data manipulation?
I currently work for a big corp with lot of tools and languages, the only common language between eng, scientists and analyst is sql. Also persistence (views or tables) could be useful and I can reuse DuckDB transformations in our DW (BigQuery) later on with few changes.

2. Do you actually have a Duck DB in disk?
I use both (memory and disk). In disk is great for having a kind of local "datamart"; in memory resembles more a typical dataframe workflow.
Also "Running on a persistent database allows spilling to disk, thus facilitating larger-than-memory workloads (i.e., out-of-core-processing)."

> 3. How are you currently importing data into Duck DB to work with it?
I usually import csv or parquet files directly to duckdb

**Comment by josevalim** [comment]
Thank you! This looks good to me. If you are happy with it, we can merge it.

**Comment by josevalim** [comment]
To be clear, let me know if it is ready by saying yes or no :)

File: lib/kino_db/connection_cell.ex [hunk 3]
```diff
@@ -227,6 +233,19 @@ defmodule KinoDB.ConnectionCell do
     end
   end
 
+  defp to_quoted(%{"type" => "duckdb"} = attrs) do
+    var = quoted_var(attrs["variable"])
+    path = quoted_var(attrs["database_path"])
+    IO.inspect(path)
```
Comment (jonatanklosko) [inline lib/kino_db/connection_cell.ex:239]:
```suggestion
```

File: lib/kino_db/connection_cell.ex [hunk 4]
```diff
@@ -159,6 +162,9 @@ defmodule KinoDB.ConnectionCell do
         "sqlite" ->
           ~w|database_path|
 
+        "duckdb" ->
+          []
```
Comment (jonatanklosko) [inline lib/kino_db/connection_cell.ex:166]: Don't we always need the path?

```suggestion
          ~w|database_path|
```

File: lib/kino_db/connection_cell.ex (hunk 3 above)
Comment (matreyes) [reply lib/kino_db/connection_cell.ex]: good catch!

File: lib/kino_db/connection_cell.ex (hunk 4 above)
Comment (matreyes) [reply lib/kino_db/connection_cell.ex:166]: It is optional, if unset it runs on memory.

**Comment by matreyes** [comment]
Thanks @jonatanklosko ,
It's OK from my side.

File: lib/kino_db/connection_cell.ex (hunk 4 above)
Comment (jonatanklosko) [reply lib/kino_db/connection_cell.ex:166]: I see, in that case I think we should generate the code without `path` if it is empty, wdyt?

Comment (matreyes) [reply lib/kino_db/connection_cell.ex:166]: sure, making it more explicit. No problem!

File: lib/kino_db/connection_cell.ex [hunk 5]
```diff
@@ -227,6 +233,23 @@ defmodule KinoDB.ConnectionCell do
...
+      adbc_db_conf =
+        case unquote(attrs["database_path"]) do
+          "" -> {Adbc.Database, driver: :duckdb}
+          path -> {Adbc.Database, driver: :duckdb, path: path}
+        end
+
+      {:ok, db} = Kino.start_child(adbc_db_conf)
+      {:ok, unquote(var)} = Kino.start_child({Adbc.Connection, database: db})
+    end
+  end
```
Comment (jonatanklosko) [inline lib/kino_db/connection_cell.ex:251]: We want the generation itself to be conditional, like this:

```suggestion
  defp to_quoted(%{"type" => "duckdb"} = attrs) do
    var = quoted_var(attrs["variable"])
    
    opts =
      if path = attrs["database_path"] do
        [driver: :duckdb, path: path]
      else
        [driver: :duckdb]
      end

    quote do
      :ok = Adbc.download_driver!(:duckdb)

      {:ok, db} = Kino.start_child({Adbc.Database, unquote(opts)})
      {:ok, unquote(var)} = Kino.start_child({Adbc.Connection, database: db})
    end
  end
```

(Please test that it works :))

Comment (matreyes) [reply lib/kino_db/connection_cell.ex]: sure, it works as expected, but attrs["database_path"] could be an empty, falling always in the first clause, I had to use case to check for empty string

Comment (jonatanklosko) [reply lib/kino_db/connection_cell.ex]: Oh yeah, that should be the check definitely!

**Review by jonatanklosko — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**